### ISSUE TITLE
Quantitative APA data for survival analyses

### ISSUE THREAD

**Issue by mradovic** [shenkers/isoscm#17]
Hello,
  I am interesting in using isoscm for some survival analyses. We have RNA-seq data from tumor samples, where we would like to get a quantitative measurement of APA (3'UTR-shortening) on a gene-by-gene basis, and then to take those quantitative values to do a survival analysis (cox regression, kaplan-meier). Would I be able to use IsoSCM to get a quantitative output to do that analysis? Thank you so much for your consideration!

Best,
Milan
mradovic@iupui.edu

**Comment by shenkers** [comment]
Hi Milan, 

The output of the 'compare' command includes a 'differential_usage' estimate for each identified polyadenylation site between a pair of samples. You could use this differential_usage estimates calculated for each gene as covariates for your cox analysis. There are some things I would consider: 
- do all the tumors use the same shortened 3'UTR isoform? I imagine the short isoforms may not be the same between different tumors.
- Do you want to group different short isoforms together as a single shortening-of-gene-X event, or treat each short isoform as a separate covariates?
- since the 3'UTR shortening events identified by the compare command will likely be different between samples, you may want to create a unified catalogue of all shortening events identified by IsoSCM across all your samples, and in a second step quantify the usage of these sites across all your samples

Best,
Sol

**Comment by mradovic** [comment]
Hi Sol,

   This is very helpful, I appreciate the thoughtful response! I agree with you, I would not expect all tumors to use the same shortened 3'UTR isoform. There will definitely be quite a bit of heterogeneity. I think your idea of creating a unified catalogue, and then quantifying each shortening event across our sample set would be the right way to go. Would you have some guidance on how best to do this? Im guessing running "assemble" across all samples, then somehow concatenating the GTF files, and then quantifying? With quantification, as the "compare" command requires a comparison to another sample, not sure what would be the best way. Thanks for your assistance!

Milan

**Comment by shenkers** [comment]
Yes, the compare command is for comparing two samples, so it would require some adaptation to apply to your analysis. 

I do have a more experimental version of the program I could share that would allow you to run the segmentation on more than two samples. In this case you would run the assemble command on all k samples independently, then run a segmentation step that looks at all the samples simultaneously to identify change points. I have this code on a separate branch that is diverged somewhat from the master version, so it may take some time to merge into the main code base. I can share it now, as is, but it will take some time before I am able to merge this feature with version 2.0.12.

Alternatively, if you want to use the current version of IsoSCM, it would take a little more work and some scripting. I think your approach sounds reasonable to assemble each sample and concatenate all the change points. If you wanted to go this route I would recommend:

1) find the longest 3'UTR isoform for each gene across all samples
2) collect all change-points that overlap this genome segment
3) aggregate change-points that occur within a short distance of each other (say, less than 250nt)
4) output all the 3'UTR isoforms by segmenting the longest 3'UTR segment from 1) with the aggregated change-points from 3)
5) use a tool like miso or dexseq to quantify differential usage of these 3'utr isoforms across your samples

**Comment by mradovic** [comment]
Hi Sol,
  Sounds good! If you would be willing to share the new segmentation code, Id be happy to give it a shot. Ill then run the assemble command on all the samples independently with the current code, and feed the output into your new code. Really appreciate the assistance!

Milan

**Comment by shenkers** [comment]
I put together a little example of how to run the experimental version. Here I have simulated 4 samples that each use 3 alternative 3'ends to a different extent:

![multi-segmentation](https://cloud.githubusercontent.com/assets/2341961/13157445/f037e7c8-d655-11e5-937e-dd05851a5e0f.png)

The 4 files are called multi.bam,multi1.bam,multi2.bam,multi3.bam, to run IsoSCM on these samples you first run the assemble command

```
for i in *.bam ; do 
    java -jar IsoSCM-exp.jar assemble -bam $i -s unstranded -base `basename $i .bam`; 
done
```

The output of this will include files matching the pattern "isoscm/*.assembly_parameters.xml", which are the input for the new joint segmentation step:

```
java -jar ../IsoSCM-2.0.6.jar joint -xml isoscm/multi.assembly_parameters.xml,isoscm/multi1.assembly_parameters.xml,isoscm/multi2.assembly_parameters.xml,isoscm/multi3.assembly_parameters.xml
```

The quantification will be in the file joint/multi.multi1.multi2.multi3.txt, here is an excerpt of the columns that you would mostly be interested in.

| samples | changepoint_id | n_reads_upstream | n_reads_downstream | site_usage |
| --- | --- | --- | --- | --- |
| multi,multi1,multi2,multi3 | 000000000 | 319,217,196,212 | 107,0,165,157 | 0.58,0.99,0.0,0.05 |
| multi,multi1,multi2,multi3 | 000000001 | 107,0,165,157 | 23,0,0,120 | 0.63,1.0,0.99,0.0 |

So the way you read this is each row corresponds to a changepoint, and the columns are various attributes of that changepoing. The first column lists the sample names, and any other column that has comma separated values, these values are listed in the same order as in this column. So the first row is the changepoint with ID:000000000 the sample 'multi' has a 58% drop in coverage at this site, sample 'multi1' has a 99% drop in coverage, and samples 'multi2' and 'multi3' have a ~0% drop in coverage. 

The second changepoint illustrates a case to keep in mind when you look at your own data. Here, you can see that this changepoint has zero reads in the upstream and downstream segments for sample 'multi1', and since the site_usage = 1-(downstream_coverage/upstream_coverage) this value is undefined when the upstream_coverage = 0 (you can't divide by zero). These cases get assigned a usage of 1 right now, but I think you should exclude usage estimates when the upstream coverage is below a threshold amount, since these estimates will not be very confident.

there is also a GTF file multi1.multi2.multi3.multi.gtf that you can load in IGV to more easily visualize the change point locations (shown at the bottom of the image above).

I have attached a zip containing the executable jar and the simulated bam files I am showing in the image above.

[IsoSCM-exp.zip](https://github.com/shenkers/isoscm/files/136909/IsoSCM-exp.zip)

**Comment by mradovic** [comment]
Thank you so much sol! The details are very helpful. Im going to try to give this a run today or this weekend, and will let you know.

**Comment by shenkers** [comment]
Of course! let me know if you run into any issues

**Comment by mradovic** [comment]
Hi Sol,
  I was successful in running “assemble” on our files. Im now working on the joint segmentation. I have 385 xml files undergoing joint segmentation. Everything runs ok for a while, but I then get "out of memory" errors. We have some large memory servers (up to 512GB of RAM per node). I can increase heap size, but ill max out what 32-bit Java can do. Do you have any thoughts on how to overcome this? Will IsoSCM run on 64-bit java, and thus I can increase heap size?

Best,
Milan

**Comment by mradovic** [comment]
Hi Sol,

  Please disregard my last message. I was able to get the memory issue figured out, and the joint completed ok. The only issue now, is that the GTF file from the joining only has 120 entries, which seems a little low. In particular given 385 samples. Ive attached the gtf file? Btw... I completely understand this is experimental, and appreciate truly your efforts to accommodate. So no worries, if we can't do this type of analysis at this time.

Best,

Milan​

**Comment by shenkers** [comment]
Hi Milan,

I'm glad you got the memory issue resolved. If you still have it, could you include the stack trace where it crashes? Just in case, I want to check if it hints at some memory leak. You should be able to run IsoSCM with a 64bit jvm and give it more memory.

About the 120 entries being small, I think this is a nuance of how the 'joint' step is performed. First it looks at all your assemblies (385 in this case), and identifies 3'-terminal exons that are common between all samples. These exons are carried forward to the segmentation step. Do your samples come from very diverse cell types? It could be that there are few terminal exons that are consistently expressed among all your samples. It sounds like requiring that the exon be represented in all samples is also probably too strict for your analysis. I guess what we want to do is perform the segmentation on the subset of samples where that exon _is_ expressed, and report missing values when it is not detected.

By the way, the GTF didn't get attached (I don't the attachments are included if you are responding directly to the github email)

**Comment by mradovic** [comment]
Hi Sol,
   I agree. The samples are all from Colon Cancers, which I’m sure there is considerable inter-sample variability. Thus, its probably too strict for the exon to be represented in all samples. Ive provided a link for the gtf file:
https://slashtmp.iu.edu/files/download?FILE=mradovic%2F508833Mfiwd

password: genomics

Best,
Milan

**Comment by mradovic** [comment]
Hi Sol,
  As I was thinking about this, would it be possible with the “joining” algorithm as it now stands, to either be able to set a threshold value for the percentage of samples that have to have a certain change point in order to be reported? Like 10% or so. Or alternatively, is there a way to disengage the requirement that the change point be present in all samples, and just get the full output? Thanks!

Best,
Milan

From: mradovic <mradovic@iupui.edu<mailto:mradovic@iupui.edu>>
Date: Friday, February 26, 2016 at 11:39 AM
To: shenkers/isoscm <reply@reply.github.com<mailto:reply@reply.github.com>>
Subject: Re: [isoscm] Quantitative APA data for survival analyses (#17)

Hi Sol,
   I agree. The samples are all from Colon Cancers, which I’m sure there is considerable inter-sample variability. Thus, its probably too strict for the exon to be represented in all samples. Ive provided a link for the gtf file:
https://slashtmp.iu.edu/files/download?FILE=mradovic%2F508833Mfiwd

password: genomics

Best,
Milan

**Comment by shenkers** [comment]
Hi Milan,
Thank you for your patience, I think these are both good ideas. I have
something I need to finish before Wednesday, but I will get back to you on
Thursday.
Best,
Sol

**Comment by mradovic** [comment]
Hi Sol,
  No problem at all. I know this is "extra", so I truly appreciate it.

Milan

Sent from my iPhone

On Feb 29, 2016, at 4:42 PM, shenkers <notifications@github.com<mailto:notifications@github.com>> wrote:

Hi Milan,
Thank you for your patience, I think these are both good ideas. I have
something I need to finish before Wednesday, but I will get back to you on
Thursday.
Best,
Sol

On Mon, Feb 29, 2016 at 4:14 PM, mradovic <notifications@github.com<mailto:notifications@github.com>> wrote:

> Hi Sol,
> As I was thinking about this, would it be possible with the "joining"
> algorithm as it now stands, to either be able to set a threshold value for
> the percentage of samples that have to have a certain change point in order
> to be reported? Like 10% or so. Or alternatively, is there a way to
> disengage the requirement that the change point be present in all samples,
> and just get the full output? Thanks!
> 
> Best,
> Milan
> 
> From: mradovic <mradovic@iupui.edu<mailto:mradovic@iupui.edu>mailto:mradovic@iupui.edu>
> Date: Friday, February 26, 2016 at 11:39 AM
> To: shenkers/isoscm <reply@reply.github.com<mailto:reply@reply.github.com>mailto:reply@reply.github.com>
> 
> Subject: Re: [isoscm] Quantitative APA data for survival analyses (#17)
> 
> Hi Sol,
> I agree. The samples are all from Colon Cancers, which I'm sure there is
> considerable inter-sample variability. Thus, its probably too strict for
> the exon to be represented in all samples. Ive provided a link for the gtf
> file:
> https://slashtmp.iu.edu/files/download?FILE=mradovic%2F508833Mfiwd
> 
> password: genomics
> 
> Best,
> Milan
> 
> From: shenkers <notifications@github.com<mailto:notifications@github.com>mailto:notifications@github.com>
> 
> Reply-To: shenkers/isoscm <reply@reply.github.com<mailto:reply@reply.github.com><mailto:
> reply@reply.github.com<mailto:reply@reply.github.com>>>
> Date: Friday, February 26, 2016 at 10:29 AM
> To: shenkers/isoscm <isoscm@noreply.github.com<mailto:isoscm@noreply.github.com><mailto:
> isoscm@noreply.github.com<mailto:isoscm@noreply.github.com>>>
> Cc: mradovic <mradovic@iupui.edu<mailto:mradovic@iupui.edu>mailto:mradovic@iupui.edu>
> Subject: Re: [isoscm] Quantitative APA data for survival analyses (#17)
> 
> Hi Milan,
> 
> I'm glad you got the memory issue resolved. If you still have it, could
> you include the stack trace where it crashes? Just in case, I want to check
> if it hints at some memory leak. You should be able to run IsoSCM with a
> 64bit jvm and give it more memory.
> 
> About the 120 entries being small, I think this is a nuance of how the
> 'joint' step is performed. First it looks at all your assemblies (385 in
> this case), and identifies 3'-terminal exons that are common between all
> samples. These exons are carried forward to the segmentation step. Do your
> samples come from very diverse cell types? It could be that there are few
> terminal exons that are consistently expressed among all your samples. It
> sounds like requiring that the exon be represented in all samples is also
> probably too strict for your analysis. I guess what we want to do is
> perform the segmentation on the subset of samples where that exon is
> expressed, and report missing values when it is not detected.
> 
> By the way, the GTF didn't get attached (I don't the attachments are
> included if you are responding directly to the github email)
> 
> ## 
> 
> Reply to this email directly or view it on GitHub<
> https://github.com/shenkers/isoscm/issues/17#issuecomment-189323066>.
> 
> ## 
> 
> Reply to this email directly or view it on GitHub
> https://github.com/shenkers/isoscm/issues/17#issuecomment-190399227.

##

**Comment by shenkers** [comment]
Hi Milan, 
I just wanted to let you know that I'm working on this, and I think I should be able to send an update tomorrow.
Sol

**Comment by mradovic** [comment]
Thank you Sol, I really appreciate it!!

Sent from my iPhone

On Mar 3, 2016, at 5:35 PM, shenkers <notifications@github.com<mailto:notifications@github.com>> wrote:

Hi Milan,
I just wanted to let you know that I'm working on this, and I think I should be able to send an update tomorrow.
Sol

##

**Comment by shenkers** [comment]
Hi Milan, I think I have made the necessary adjustments. In the previous version, if a terminal exon was not represented in at least one sample it was excluded from the segmentation step. In the new version I've modified the program so that now it jointly segments all samples for which a compatible 3'exon was constructed during the assemble step. 

The instructions to run the program are the same as before, but the output will probably include more than 120 entries. 

The format of the output table is basically the same, however now the number of entries for each changepoint can be variable, if that 3'exon is not expressed in all samples that are being analyzed. As before the 'sample' column will tell you which samples the changepoint statistics correspond to.

I haven't tested this very extensively, so let me know if you run into any problems.

[isoscm-160304.zip](https://github.com/shenkers/isoscm/files/159212/isoscm-160304.zip)

**Comment by mradovic** [comment]
Thank you Sol, I truly appreciate it! Ill give this a shot this weekend, and let you know how it goes.

Best,
Milan

**Comment by mradovic** [comment]
Hi Sol,
  Just wanted to let you know that I ran the new joint algorithm, and it just completed. It took 133 hours, but it finished! With 385 samples, the gtf has 530,018 lines. Of that 209,184 lines are change points. Im guessing many of these are spurious, and may only be present in one sample. Any thoughts on how I could filter this? Im guessing maybe a command line where I can script for it occurring in 10% (39 samples) or more. Again, thank you so much for your fantastic help!

If you would like, the gtf file can be downloaded here: https://slashtmp.iu.edu/files/download?FILE=mradovic%2F77008JMWijc
password: genomics

Milan

**Comment by shenkers** [comment]
Hi Milan,

I'm glad to hear the update, I've never had this many samples to run it on simultaneously, so I was curious how it would go.

To filter for changepoints I would recommend the following criteria, in addition to a changepoint being reproducible:
- The up/downstream segments are sufficiently long (ie, 150nt). If the data is noisy, a local fluctuation in coverage will have a bigger impact on the estimated usage if the intervals are very short
- A minimum number of reads upstream of the changepoint (ie 50 reads). The more highly expressed a gene is the more accurately the usage can be estimated
- A threshold on the differential usage (i.e. 20-50%). Large differential usage values are less likely to reflect random fluctuations in coverage between samples

I typically filter the .txt file output, would it be possible to share that as well? I can write a command line script to filter the output for you.

Sol

**Comment by mradovic** [comment]
Hi Sol,
  FantasticI The link for the txt file is below. I sent the gtf, because that what ill need to do quantification with another software, but if the txt file could be used also, that ok too.

Best,
Milan

https://slashtmp.iu.edu/files/download?FILE=mradovic%2F62935Wl8fZ0
password: genomics

**Comment by shenkers** [comment]
Hi Milan,

Here's a first pass script to filter the output, it filters the output to select events where both the short and long isoform are represented in a minimum number of samples (at least 2 in the script below), so these should represent truncations that occur recurrently among samples. It was unclear if you have matched normal tissue, but if you do then it could be useful to look at shortening events that occur in tumor relative to the control samples.

There are a couple parameters you may want to adjust that are listed at the top of the file

length_thresh - the minimum length of the segments up and downstream of the changepoint, now 150nt
min_reads - the minmum number of reads upstream of a changepoint for the estimated usage to be considered, now 50 reads
min_samples - number of samples that meat the minimum number of reads, right now 39
max_usage - usage threshold for a site to be considered used, right now 80%
min_usage - usage threshold for a site to be considered skipped, right now 20%
min_number_using - number of samples above the max_usage, right now 2
min_number_skipping - number of samples below the min_usage, 2

The output of the script will be a table with three columns, the locus_id, the number of samples using the site, and the number of samples skipping the site. 

``` bash
#!/bin/bash

file=$1

length_thresh=150
min_reads=50
min_samples=39
max_usage=0.8
min_usage=0.2
min_number_using=2
min_number_skipping=2

echo -e "locus_id\tn_using\tn_skipping"
awk '{ if(NR>1) print }' $file | \
while read samples locus_id changepoint_id changepoint confidence log_odds upstream_segment downstream_segment locus strand n_reads_upstream n_reads_downstream upstream_cov downstream_cov site_usage; do
    ucoord=(`echo $upstream_segment | sed 's/:\|-/ /g'`)
    dcoord=(`echo $downstream_segment | sed 's/:\|-/ /g'`)
    ulen=`echo ${ucoord[2]} - ${ucoord[1]} + 1 | bc`
    dlen=`echo ${dcoord[2]} - ${dcoord[1]} + 1 | bc`

    # make sure the up and downstream segments are long enough
    if (( $ulen < $length_thresh )); then
        continue;
    fi
    if (( $dlen < $length_thresh )); then
        continue;
    fi

    indicesWithSufficientReads=(`
    idx=0; \
    echo $n_reads_upstream | \
    sed 's/,/\n/g' | \
    while read i; do 
        if (( $((i+1)) > $min_reads )); then 
            echo $idx; 
        fi; 
        idx=$((idx+1))
    done `)

    # make sure this site has enough reads in enough samples
    if (( ${#indicesWithSufficientReads[@]} < $min_samples )); then
        continue;
    fi

    nAboveMaxUsage=0
    nBelowMinUsage=0
    usageWithSufficientReads=();
    usages=(${site_usage//,/ })
    for idx in ${indicesWithSufficientReads[@]};do
        usageWithSufficientReads+=(${usages[$idx]})
    done
    for usage in ${usageWithSufficientReads[@]}; do
        if (( $( echo "$usage > $max_usage" | bc -l ) )); then
            nAboveMaxUsage=$((nAboveMaxUsage + 1))
        fi
        if (( $( echo "$usage < $min_usage" | bc -l ) )); then
            nBelowMinUsage=$((nBelowMinUsage + 1))
        fi
    done

    # check if enough samples use or skip this site
    if (( $nAboveMaxUsage < $min_number_using )); then
        continue;
    fi
    if (( $nBelowMinUsage < $min_number_skipping )); then
        continue;
    fi
    echo -e "$locus_id\t$nBelowMinUsage\t$nAboveMaxUsage"
done
```

if you put this in a file called script.sh, you would run it like:

``` bash
./script.sh COAD_joint.txt > out.txt
# to pull out the lines that pertaining to the filtered changepoint ids you can use 'join'
# the -t parameter indicates the fields are tab separated, to 
# enter a tab character on the command line you need to type
# ctrl-v then let go and press <tab>
join -t'   ' -1 3 <( tail -n+2 COAD_joint.txt | sort -k3,3 ) <( tail -n+2 out.txt | cut -f 1 )
```

For example, it looks like changepoint 000001350 at chr20:61576941-61576941 is used in 8 out of 142 samples meeting the expression criteria, so it may be a recurrent shortening event. It may be interesting to check if specific samples are more prone to express shortened 3'UTRs, which if present might indicate there is a general disregulation of polyadenylation in these samples. 

I don't have your original data, so it's difficult for me to spot check this, you may need to adjust the parameters if these values give results that are too stringent/lenient when you browse these loci. Also, it's a little slow using a shell script, but if you don't mind waiting, it should do the trick.

**Comment by mradovic** [comment]
Thanks Sol, this is fantastic! Ill try to give it a whack here shortly.
